fs: zms: multiple fixes from previous PR review (#80407)
zephyrproject-rtos/zephyr: doc/services/storage/zms/zms.rst

		};
	} __packed;

.. note:: The CRC of the data is checked only when the whole the element is read.
   The CRC of the data is not checked for a partial read, as it is computed for the whole element.

.. note:: Enabling the CRC feature on previously existing ZMS content without CRC enabled
   will make all existing data invalid.

.. _free-space:

Available space for user data (key-value pairs)
***********************************************

> For both scenarios ZMS should always have an empty sector to be able to perform the

Author: That was referring to deleted lines, will remove that

garbage collection (GC).
So, if we suppose that 4 sectors exist in a partition, ZMS will only use 3 sectors to store
Key-value pairs and keep one sector empty to be able to launch GC.
The empty sector will rotate between the 4 sectors in the partition.

.. note:: The maximum single data length that could be written at once in a sector is 64K
   (This could change in future versions of ZMS)

Small data values
=================

> Values smaller than 8 bytes will be stored within the entry (ATE) itself, without writing data

Author: entry's ATE is a repetition. ATE is the entry

=================

Large data values ( > 8 bytes) are stored separately at the top of the sector.
In this case, it is hard to estimate the free available space, as this depends on the size of

> the data. But we can take into account that for N bytes of data (N > 8 bytes) an additional

Author: small data values ( <= 8 bytes) will always occupy the same memory space (16 bytes)

Using storage systems that rely on an erase-value (NVS as an example) will need to emulate the
erase with write operations. This will cause a significant decrease in the life expectancy of
these devices and will cause more delays for write operations and for initialization.
ZMS uses a cycle count mechanism that avoids emulating erase operation for these devices.
It also guarantees that every memory location is written only once for each cycle of sector write.

As an example, to erase a 4096 bytes sector on a non-erasable device using NVS, 256 flash writes
must be performed (supposing that write-block-size=16 bytes), while using ZMS only 1 write of
16 bytes is needed. This operation is 256 times faster in this case.

Garbage collection operation is also adding some writes to the memory cell life expectancy as it
is moving some blocks from one sector to another.
To make the garbage collector not affect the life expectancy of the device it is recommended

> to correctly dimension the partition size. Its size should be the double of the maximum size of

Author: To size, ...  Its size => In my opinion it is a repetition. I prefer the first expression

both solutions could be implemented, the best choice should be based on the calculations of the
life expectancy of the device described in this section: :ref:`wear-leveling`.

Recommendations to increase performance
***************************************

Sector size and count
=====================

> - The total size of the storage partition should be well dimensioned to achieve the best

Reviewer: As I've already commented elsewhere, I don't think `dimensioned` is the best choice here. e.g. simpler alternative here (either-or):
```suggestion
- The total size of the storage partition should be set carefully/appropriately to achieve the best
```

>   performance for ZMS.

Reviewer: ```suggestion
  performance with ZMS.
```

  All the information regarding the effectively available free space in ZMS can be found
  in the documentation. See :ref:`free-space`.

>   We recommend choosing a storage partition that can hold double the size of the key-value pairs
>   that will be written in the storage.

Reviewer: ```suggestion
  It's recommended to choose a storage partition size that is double the size of the key-value pairs
  that will be written in the storage.
```
Do we have to count the key sizes (32 bits) in that? It's a bit unclear.

Author: Actually we have to count the total size of the key-value pair. whether it is a small data value (16 bytes for each key/value pair) or large data value (16 bytes + sizeof(data))

> - The size of a sector needs to be dimensioned to hold the maximum data length that will be stored.

Reviewer: ```suggestion
- The sector size needs to be set to be able to hold the maximum data size that will be stored.
```

>   Increasing the size of a sector will slow down the garbage collection operation which will
>   occur less frequently.

Reviewer: ```suggestion
  Increasing the sector size will slow down the garbage collection operation and make it occur less frequently.
```

>   Decreasing its size, in the opposite, will make the garbage collection operation faster
>   which will occur more frequently.

Reviewer: ```suggestion
  Decreasing its size, on the opposite, will make the garbage collection operation faster but also occur more frequently.
```

> - For some subsystems like :ref:`Settings <settings_api>`, all path-value pairs are split into two ZMS entries (ATEs).

Reviewer: Why `some subsystems like`? Isn't it only about settings?

Author: ZMS could be used directly by any subsystem (bluetooth, secure storage, ...)
This is an example of one subsystem which is Settings

>   The header needed by the two entries should be accounted when computing the needed storage space.

Reviewer: ```suggestion
  The headers needed by the two entries should be accounted for when computing the needed storage space.
```

> - Using small data to store in the ZMS entries can increase the performance, as this data is

Reviewer: ```suggestion
- Storing small data (<= 8 bytes) in ZMS entries can increase the performance, as this data is
```

  written within the entry header.
  For example, for the :ref:`Settings <settings_api>` subsystem, choosing a path name that is
  less than or equal to 8 bytes can make reads and writes faster.

> Dimensioning cache

Reviewer: ```suggestion
Cache size
```

==================

> - When using ZMS API directly, the recommended cache size should be, at least, equal to
>   the number of different entries that will be written in the storage.

Reviewer: ```suggestion
- When using the ZMS API directly, the recommendation for the cache size is to make it at least equal to
  the number of different entries that will be written in the storage.
```
And same for the second occurrence of that phrasing below.

- Each additional cache entry will add 8 bytes to your RAM usage. Cache size should be carefully
  chosen.
- If you use ZMS through :ref:`Settings <settings_api>`, you have to take into account that each Settings entry is

>   divided into two ZMS entries. The perfect cache size should be double that of the Settings

Author: you are right, the cache is using a random hash function
